### PR TITLE
[patch] divide gateway for future architecture

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <i.can.feel.gravity@gmail.com>

### Description:

This PR includes a lot of feature.
### Environment:

- Go Version: 1.14.4
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.0

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kevindiu** [comment]
I think it should be `minor` or `major` release 🤔

File: internal/core/ngt/ngt.go [hunk 1]
```diff
@@ -97,7 +97,7 @@ type (
 // ObjectType is alias of object type in NGT
 type objectType int
 
-// DistanceType is alias of distance type in NGT
+// DistanceType is alias of distance type in NGTErrInvalidVector
```
Comment (rinx) [inline internal/core/ngt/ngt.go:100]: badge
by accidentally?

**Comment by rinx** [comment]
badge
In this PR, search, insert, update operations in core/agent.proto are removed and a service in gatway/vald/vald.proto is implemented to agent-NGT.
however, before reviewing this PR, I imagined that restructuring of proto files would be done by:

- remove gateway/vald/vald.proto
- remove search, insert, update operations in core/agent.proto.
- add primitive services about ANNs
    - InsertService
        - provides Insert, StreamInsert, MultiInsert
    - SearchService
        - provides Search, StreamSearch, SearchByID, ...
    - UpdateService
    - DeleteService

and these services are implemented to agent-NGT and all new gateways.
i think it's more flexible than the current definition.

and in the current implementation, agent-NGT implements the Vald service that is defined in gateway/vald/vald.proto.
i think it's a little bit strange that "agent has gateway's service". it should be "gateway has agent's Insert, Search, Update, and Delete services".

How do you think about it?

File: pkg/gateway/filter/handler/grpc/handler.go [hunk 2]
```diff
@@ -0,0 +1,587 @@
...
+
+func (s *server) Update(ctx context.Context, vec *payload.Object_Vector) (res *payload.Object_Location, err error) {
+	ctx, span := trace.StartSpan(ctx, apiName+".Update")
+	defer func() {
+		if span != nil {
+			span.End()
+		}
+	}()
+	meta := vec.GetId()
+	uuid, err := s.metadata.GetUUID(ctx, meta)
```
Comment (hlts2) [inline pkg/gateway/filter/handler/grpc/handler.go:277]: Please teach me about the following lines. :bow:
From my own understanding, I thought that filter does not connect to the Vald Meta, but only proxies to the underlying gateway (Vald MetaGateway). 🤔

File: pkg/gateway/filter/handler/grpc/handler.go [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+	vald.ValdServer
+	filter.FilterServer
+}
+
+type server struct {
+	eg                errgroup.Group
+	filter            service.Filter
+	gateway           client.Client
+	copts             []grpc.CallOption
+	streamConcurrency int
```
Comment (vankichi) [inline pkg/gateway/filter/handler/grpc/handler.go:51]: there is no `metadata` in server struct.
please confirm it for all service handler file 🙏

File: Makefile [hunk 4]
```diff
@@ -22,16 +22,21 @@ BASE_IMAGE                      = $(NAME)-base
...
 AGENT_SIDECAR_IMAGE             = $(NAME)-agent-sidecar
 GATEWAY_IMAGE                   = $(NAME)-gateway
+BACKUP_GATEWAY_IMAGE            = $(NAME)-backup-gateway
+CI_CONTAINER_IMAGE              = $(NAME)-ci-container
 DISCOVERER_IMAGE                = $(NAME)-discoverer-k8s
-META_REDIS_IMAGE                = $(NAME)-meta-redis
-META_CASSANDRA_IMAGE            = $(NAME)-meta-cassandra
-MANAGER_BACKUP_MYSQL_IMAGE      = $(NAME)-manager-backup-mysql
+FILTER_GATEWAY_IMAGE            = $(NAME)-filter-gateway
+GATEWAY_IMAGE                   = $(NAME)-gateway
```
[the remark this answers is not in the document]
Comment (kevindiu) [reply Makefile:29]: I think type and then alphabetical order is better :)

File: pkg/gateway/filter/handler/grpc/handler.go (hunk 2 above)
Comment (kpango) [reply pkg/gateway/filter/handler/grpc/handler.go:277]: filter gateway is another task please ignore them.

File: pkg/gateway/filter/handler/grpc/handler.go (hunk 3 above)
Comment (kpango) [reply pkg/gateway/filter/handler/grpc/handler.go:51]: filter gateway is another task please ignore them.

File: apis/proto/vald/insert.proto [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+//
+
+syntax = "proto3";
+
+package vald;
+
+option go_package = "github.com/vdaas/vald/apis/grpc/vald";
+option java_multiple_files = true;
+option java_package = "org.vdaas.vald";
+option java_outer_classname = "ValdApi";
```
Comment (rinx) [inline apis/proto/vald/insert.proto:24]: badge
Could you please modify the Java package and class name correctly?
I recommend to modify like this:
```suggestion
option java_package = "org.vdaas.vald.insert";
option java_outer_classname = "InsertApi";
```

File: apis/proto/vald/object.proto [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+//
+
+syntax = "proto3";
+
+package vald;
+
+option go_package = "github.com/vdaas/vald/apis/grpc/vald";
+option java_multiple_files = true;
+option java_package = "org.vdaas.vald";
+option java_outer_classname = "ValdApi";
```
Comment (rinx) [inline apis/proto/vald/object.proto:24]: same.

File: apis/proto/vald/remove.proto [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+//
+
+syntax = "proto3";
+
+package vald;
+
+option go_package = "github.com/vdaas/vald/apis/grpc/vald";
+option java_multiple_files = true;
+option java_package = "org.vdaas.vald";
+option java_outer_classname = "ValdApi";
```
Comment (rinx) [inline apis/proto/vald/remove.proto:24]: same.

File: apis/proto/vald/search.proto [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+//
+
+syntax = "proto3";
+
+package vald;
+
+option go_package = "github.com/vdaas/vald/apis/grpc/vald";
+option java_multiple_files = true;
+option java_package = "org.vdaas.vald";
+option java_outer_classname = "ValdApi";
```
Comment (rinx) [inline apis/proto/vald/search.proto:24]: same.

File: apis/proto/vald/update.proto [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+//
+
+syntax = "proto3";
+
+package vald;
+
+option go_package = "github.com/vdaas/vald/apis/grpc/vald";
+option java_multiple_files = true;
+option java_package = "org.vdaas.vald";
+option java_outer_classname = "ValdApi";
```
Comment (rinx) [inline apis/proto/vald/update.proto:24]: same.

File: apis/proto/vald/upsert.proto [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+//
+
+syntax = "proto3";
+
+package vald;
+
+option go_package = "github.com/vdaas/vald/apis/grpc/vald";
+option java_multiple_files = true;
+option java_package = "org.vdaas.vald";
+option java_outer_classname = "ValdApi";
```
Comment (rinx) [inline apis/proto/vald/upsert.proto:24]: same.

File: apis/proto/vald/insert.proto (hunk 5 above)
Comment (kpango) [reply apis/proto/vald/insert.proto:24]: I agree with classname change but do we need change package name?

Comment (rinx) [reply apis/proto/vald/insert.proto:24]: It is okay not to change, but `org.vdaas.vald` is already used as root package in other protos. so it is better to make a new package like `org.vdaas.vald.xxx`.

Comment (rinx) [reply apis/proto/vald/insert.proto:24]: how about to use `org.vdaas.vald.vald`?

Comment (kpango) [reply apis/proto/vald/insert.proto:24]: How about `org.vdaas.vald.api` ?

Comment (rinx) [reply apis/proto/vald/insert.proto:24]: sounds good :+1:

**Comment by rinx** [comment]
this PR is closed because the #740, #826 has same feature.

**Pull request closed without merging**